apps/astarte_data_updater_plant/test/astarte_data_updater_plant/data_updater/time_based_actions_test.exs

```diff
@@ -0,0 +1,1016 @@
+         %{
+           realm: realm,
+           encoded_device_id: encoded_device_id,
+           device_id: device_id
+         } do
+      # Insert initial device state
+      state = setup_device_state(realm, device_id, encoded_device_id, [])
+
+      assert state.last_deletion_in_progress_refresh == @timestamp_us_x_10
+      assert state.discard_messages == false
+
+      Database.insert_deletion_in_progress(device_id, realm)
+
+      # Not found means it was already disconnected, succeed anyway
+      Astarte.DataUpdaterPlant.RPC.VMQPlugin.ClientMock
+      |> expect(:delete, fn data ->
+        assert %{realm_name: ^realm, device_id: ^encoded_device_id} = data
+
+        {:error, :not_found}
+      end)
+
+      new_state =
+        TimeBasedActions.reload_device_deletion_status_on_expiry(
+          state,
+          @timestamp2_us_x_10
+        )
+
+      assert new_state.last_deletion_in_progress_refresh == @timestamp2_us_x_10
+      assert new_state.discard_messages == true
+      assert new_state.connected == false
+    end
+
+    test "raises MatchError if deletion returns error", %{
+      realm: realm,
+      encoded_device_id: encoded_device_id,
+      device_id: device_id
+    } do
+      # Insert initial device state
+      state = setup_device_state(realm, device_id, encoded_device_id, [])
+
+      assert state.last_deletion_in_progress_refresh == @timestamp_us_x_10
+      assert state.discard_messages == false
+      assert state.connected == true
+
+      Database.insert_deletion_in_progress(device_id, realm)
+
+      # Simulate VMQPlugin.delete returning an error
+      Astarte.DataUpdaterPlant.RPC.VMQPlugin.ClientMock
+      |> expect(:delete, fn data ->
+        assert %{realm_name: ^realm, device_id: ^encoded_device_id} = data
+        {:error, :bad_request}
+      end)
+
+      assert_raise MatchError, fn ->
+        TimeBasedActions.reload_device_deletion_status_on_expiry(
+          state,
+          @timestamp2_us_x_10
+        )
+      end
+    end
+
+    test "does not start device deletion if check_device_deletion_in_progress returns error", %{
+      realm: realm,
+      device_id: device_id,
+      encoded_device_id: encoded_device_id
+    } do
+      # Insert initial device state
+      state = setup_device_state(realm, device_id, encoded_device_id, [])
+
+      # Simulate error scenario: update the state to use the nonexistent realm
+      nonexistent_realm_ = "nonexistent_realm_#{System.unique_integer([:positive])}"
+      state = %{state | realm: nonexistent_realm_}
+
+      log =
+        ExUnit.CaptureLog.capture_log(fn ->
+          TimeBasedActions.reload_device_deletion_status_on_expiry(
+            state,
+            @timestamp2_us_x_10
+          )
+        end)
+
+      assert log =~ "Cannot check device deletion status for"
+      assert log =~ "should_start_device_deletion"
+      assert log =~ inspect(device_id)
```

Comment:
nit: since log messages can change and this part is not crucial to the test it can be removed